pubnub-hyper/src/transport/hyper/history.rs

```diff
@@ -0,0 +1,245 @@
+        let request::DeleteHistory {
+            channels,
+            start,
+            end,
+        } = request;
+
+        // Prepare the URL.
+        let path_and_query =
+            UriTemplate::new("/v3/history/sub-key/{sub_key}/channel/{channels}{?start,end}")
+                .set_scalar("sub_key", self.subscribe_key.clone())
+                .set_list("channels", channels)
+                .set_optional_scalar("start", start)
+                .set_optional_scalar("end", end)
+                .build();
+        let url = build_uri(&self, &path_and_query)?;
+
+        // Prepare the request.
+        let req = Request::builder()
+            .method(Method::DELETE)
+            .uri(url)
+            .body(Body::empty())?;
+
+        // Send network request.
+        let response = self.http_client.request(req).await?;
+        let _data_json = handle_history_response(response).await?;
+
+        // Parse response.
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl TransportService<request::MessageCountsWithTimetoken> for Hyper {
+    type Response = response::MessageCountsWithTimetoken;
+    type Error = error::Error;
+
+    async fn call(
+        &self,
+        request: request::MessageCountsWithTimetoken,
+    ) -> Result<Self::Response, Self::Error> {
+        let request::MessageCountsWithTimetoken {
+            channels,
+            timetoken,
+        } = request;
+
+        // Prepare the URL.
+        let path_and_query =
+            UriTemplate::new("/v3/history/sub-key/{sub_key}/message-counts/{channels}{?timetoken}")
+                .set_scalar("sub_key", self.subscribe_key.clone())
+                .set_list("channels", channels)
+                .set_scalar("timetoken", timetoken)
+                .build();
+        let url = build_uri(&self, &path_and_query)?;
+
+        // Send network request.
+        let response = self.http_client.get(url).await?;
+        let data_json = handle_history_response(response).await?;
+
+        // Parse response.
+        let channels = parse_message_counts(&data_json)
+            .ok_or_else(|| error::Error::UnexpectedResponseSchema(data_json))?;
+        Ok(channels)
+    }
+}
+
+#[async_trait]
+impl TransportService<request::MessageCountsWithChannelTimetokens> for Hyper {
+    type Response = response::MessageCountsWithChannelTimetokens;
+    type Error = error::Error;
+
+    async fn call(
+        &self,
+        request: request::MessageCountsWithChannelTimetokens,
+    ) -> Result<Self::Response, Self::Error> {
+        let request::MessageCountsWithChannelTimetokens { channels } = request;
+
+        // Tranform input to the proper format.
+        let cap = channels.len();
+        let mut names = Vec::with_capacity(cap);
+        let mut timetokens = Vec::with_capacity(cap);
+        for (channel, timetoken) in channels.into_iter() {
+            names.push(channel);
+            timetokens.push(timetoken);
+        }
```

Comment:
Ah! This can be made simpler with a little Functional love.

```suggestion
        // Unzip Vector-of-Tuples into separate Vectors.
        let (names, timetokens): (Vec<_>, Vec<_>) = channels.into_iter().unzip();
```

Also works with `HashMap`